src/CompiledMatcher.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\ConstraintInterface;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompiledMatcher
+{
+    private $cacheDir;
+
+    private $checked;
+
+    public function __construct($cacheDir = null)
+    {
+        $this->cacheDir = $cacheDir ? $cacheDir : \sys_get_temp_dir();
```

Comment:
I'd at least use `sys_get_temp_dir().'/composer-constraints'` or something to avoid creating thousands of files in /tmp

Reply:
I prefixed it in cacheKey, but your idea of doing it in constructor is better. 
Fixed